platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,447 @@
+
+The pseudocode for the startup procedure will look as follows:
+
+```java
+/**
+ * Start the platform.
+ * @param state an initial state. The caller either loads it from disk if the node has run before,
+ *              or constructs a new empty state and stores the genesis roster in there
+ *              as the current active roster.
+ */
+void startPlatform(final State state) {
+    if (the State has Candidate Roster) {
+        if (isSoftwareUpgrade) {
+            // This MUST be performed under `isSoftwareUpgrade` to ensure that
+            // the entire network is being restarted, and so every node adopts the Candidate Roster,
+            // and hence no ISSes happen.
+            // Modify the state and put Candidate Roster into Active Roster, effectively clearing the Candidate Roster.
+            makeCRtheAR();
+            // May make a record of the previous Active Roster if necessary (e.g. for PCES replay)
+        }
+    }
+
+    if (the State has no Active Roster) {
+        // This should never happen, but we have to check this because we're given a state as an argument.
+        throwFatalErrorAndShutdown();
+    }
+
+    // At this point the Active Roster in the state is what we'll be using as a roster.
+}
+```
+
+### Roster changes needed for Components
+
+#### Block Proof
+
+In a fully Dynamic Address Book (DAB) where, say, the roster for round 101 might be different from the roster for round
+102, there is a need to track what roster is used for each non-ancient round.
+We propose the introduction of a queue of roster hashes stored in the state, where the index of a roster hash determines
+what round the roster is active for as shown.
+![](TSS%20Roster%20Lifecycle-Roster%20and%20Rounds.drawio.svg)
+The size of this queue would be ancient round window + roster offset.
+The roster offset refers to the number of rounds in advance that the network determines and declares the active roster
+for a specific future round.
+This acts as a proactive mechanism that allows the network to prepare for upcoming changes in the roster and ensure a
+valid transition to a future roster.
+When each round is handled, we will pop off the active roster hash for the round that just became ancient, and add a new
+active roster hash for a new future round.
+
+Before full DAB (DAB Phase 2.5), we propose to store the current and previous active rosters in this newly introduced
+queue of hashes.
+In the DAB Phase 2.5 implementation, this queue will contain only 2 hashes, with the first hash representing
+the previous active roster, and the second hash representing the current active roster.
+At upgrade boundaries, we will pop off the previous active roster, and add a new roster to the end of the queue, which
+is the new active roster.
+
+This approach provides the benefit of introducing the necessary data structure for full DAB, while also providing a
+mechanism for tracking the active roster for each round in the interim.
```

Comment:
If this queue exists, then it already provides a reference to the active roster, making a pointer to the active roster in the platform state as described above unnecessary. Should we remove it from the platform state and only store a pointer to the candidate roster (or null if none) in there?

Reply:
I am working on this bit, it's a little fuzzy at the moment. I'll clarify today.

Reply:
Actually, this is different. The map is Map<Hashes,Roster>, which is stored in the State. The map is unaware of which roster is active or otherwise. The Roster itself does not carry this information either. We still need a mechanism to track which of the stored rosters is active. That mechanism is what is addressed by introducing a queue of 2 items (roster hashes) whose tail is the active roster hash. So, to get the active roster, we retrieve the tail of the queue, and look it up in the map.

Reply:
Right. I'm not talking about the map - the map should remain as it is described. I'm talking about the HashObject pointers in the platform state. While a pointer to the candidate roster should be there, the pointers to the active and previous rosters seem redundant because they'll be present in this queue as well.

Reply:
As per the architecture review last friday, this has been removed. We now keep a pointer to only the candidate roster's hash value.

Reply:
Sounds good. Thanks!